backend/FwLite/MiniLcm/SyncHelpers/DiffCollection.cs

```diff
@@ -138,4 +221,120 @@ public static async Task<int> Diff<T>(
+        {
+            if (stable.Contains(current[j].Id))
+            {
+                beforeEntity = current[j];
+                break;
+            }
+        }
+        for (var j = i + 1; j < current.Count; j++)
+        {
+            if (stable.Contains(current[j].Id))
+            {
+                afterEntity = current[j];
+                break;
+            }
+        }
+        return new BetweenPosition
+        {
+            Before = beforeEntity?.Id,
+            After = afterEntity?.Id
+        };
+    }
+
+    private static IEnumerable<PositionDiff> DiffPositions<T, TId>(
+        IList<T> before,
+        IList<T> after,
+        Func<T, TId> identity)
+    {
+        var beforeJson = new JsonArray(before.Select(item => JsonValue.Create(identity(item))).ToArray());
+        var afterJson = new JsonArray(after.Select(item => JsonValue.Create(identity(item))).ToArray());
+
+        if (JsonDiffPatcher.Diff(beforeJson, afterJson) is not JsonObject result)
+        {
+            yield break; // no changes
+        }
+
+        foreach (var prop in result!)
+        {
+            if (prop.Key == "_t") // diff type
+            {
+                if (prop.Value!.ToString() != "a") // we're only using the library for diffing shallow arrays
+                {
+                    throw new InvalidOperationException("Only array diff results are supported");
+                }
+                continue;
+            }
+            else if (prop.Key.StartsWith("_")) // e.g _4 => the key represents an old index (removed or moved)
+            {
+                var previousIndex = int.Parse(prop.Key[1..]);
+                var delta = prop.Value!.AsArray();
+                var wasMoved = delta[2]!.GetValue<int>() == 3; // 3 is magic number for a move operation
+                int? newIndex = wasMoved ? delta[1]!.GetValue<int>() : null; // if it was moved, the new index is at index 1
+                if (newIndex is not null)
+                {
+                    yield return new PositionDiff { From = previousIndex, To = newIndex }; // move
+                }
+                else
+                {
+                    yield return new PositionDiff { From = previousIndex }; // remove
+                }
+            }
+            else // e.g. 4 => the key represents a new index
+            {
+                var addIndex = int.Parse(prop.Key);
+                yield return new PositionDiff { To = addIndex }; // add
+            }
+        }
+
+    }
+
+    private class PositionDiff
+    {
+        public int? From { get; init; }
+        public int? To { get; init; }
+    }
+}
+
+public class BetweenPosition : IEquatable<BetweenPosition>
+{
+    // The ID of the item before the new item (not the ID of the item that the new item should be before)
+    public Guid? Before { get; set; }
+    // The ID of the item after the new item (not the ID of the item that the new item should be after)
+    public Guid? After { get; set; }
```

Comment:
I agonised a bit over what this class and these properties should be called as well as if the class should even exist or we should just use a tuple.

Before and After can be really confusing, because there are 2 intuitive and contradictory meanings.

Maybe Left and Right?